File: src/DXFViewerDX/DXFViewer/DXF/ShapeGroup.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Runtime.InteropServices;

using SharpDX;
using SharpDX.Direct2D1;
using SharpDX.DXGI;

namespace DXFViewer
{
    public class ShapeGroup : DXFDotNet.ShapeGroup, IDrawableShape
    {
        protected DXFViewer.IDrawableShape mDrawableShape = null;
        protected EditBox m_editBox = null;

        protected SharpDX.Direct2D1.Bitmap mDrawBitmap = null;
        protected SharpDX.RectangleF mImageRect = new SharpDX.RectangleF();

        public ShapeGroup() : base()
        {
        }

        public ShapeGroup(DXFDotNet.ShapeGroupOption option) : base(option)
        {            
        }

        public DXFDotNet.Shape GetShapeObject()
        {
            return this;
        }

        public override DXFDotNet.ShapeGroup CreateShapeGroup()
        {
            ShapeGroup arc = new ShapeGroup();
            return arc;
        }

        public bool CreateDXResource()
        {
            if (m_pOwnLayer == null || m_pOwnLayer.Owner == null)
                return false;
            DXFViewer.DXFControl ctrl = (DXFViewer.DXFControl)m_pOwnLayer.Owner;
            SharpDX.Direct2D1.RenderTarget g = ctrl.RenderTarget;
            
            if( this.DrawingType == DrawType.IMAGE)
            {
                mDrawBitmap = ShapeGroup.LoadFromImage(g, m_img);

            }

            m_editBox = new EditBox(ctrl);
            m_editBox.CreateDXResource();
            return true;
        }

        public bool DiscardDXResource()
        {
            if (m_editBox != null)
            {
                m_editBox.DiscardDXResource();
                m_editBox = null;
            }
            if (mDrawBitmap != null)
            {
                mDrawBitmap.Dispose();
                mDrawBitmap = null;
            }
            return true;
        }
        
        protected bool GetImageSize(ref float rWidth, ref float rHeight)
        {
            if (m_pOwnLayer == null || m_pOwnLayer.Owner == null)
                return false;

            DXFViewer.DXFControl ctrl = (DXFViewer.DXFControl)m_pOwnLayer.Owner;

	        if (m_img != null)
	        {
		        if (m_imgSize.Width < 0 || m_imgSize.Height < 0)
		        {
			        rWidth = m_img.Width * m_fImageScale;
			        rHeight = m_img.Height * m_fImageScale;
		        }
		        else
		        {
			        rWidth = m_imgSize.Width * m_fImageScale;
			        rHeight = m_imgSize.Height * m_fImageScale;
		        }

                if (ctrl.DownToTop())
			        rHeight = -rHeight;

		        return true;
	        }
	        return false;
        }

        public bool Draw(SharpDX.Direct2D1.RenderTarget g, bool bDrawText)
        {
            if (m_drawType == DrawType.IMAGE)
	        {
		        if (m_img != null)
		        {
			        float fWidth = 0, fHeight = 0;
			        GetImageSize(ref fWidth, ref fHeight);
                    
                    mImageRect.X = (float)m_vPos.x;
                    mImageRect.Y = (float)m_vPos.x;
                    mImageRect.Width = fWidth;
                    mImageRect.Height = fHeight;

                    g.DrawBitmap(mDrawBitmap, mImageRect, 1.0f, BitmapInterpolationMode.Linear);

			        if (Selectable && Selected)
				        m_editBox.Draw(g, (float)(m_vPos.x + fWidth / 2), (float)(m_vPos.y + fHeight / 2));

			        return true;
		        }
	        }
	        else if (m_drawType == DrawType.SHAPE)
	        {
                if (m_shape != null)
		        {
			        m_shape.Selectable = this.Selectable;
			        m_shape.Selected = this.Selected;
			        
                    mDrawableShape = (IDrawableShape)m_shape;
                    if(mDrawableShape != null)
                    {
                        mDrawableShape.Draw(g, bDrawText);
                    }
			        return true;
		        }
	        }
            return true;
        }

        public static Bitmap LoadFromImage(RenderTarget renderTarget, System.Drawing.Image img)
        {
            // Loads from file using System.Drawing.Image
            using (var bitmap = (System.Drawing.Bitmap)new System.Drawing.Bitmap(img))
            {
                var sourceArea = new System.Drawing.Rectangle(0, 0, bitmap.Width, bitmap.Height);
                var bitmapProperties = new BitmapProperties(new PixelFormat(Format.R8G8B8A8_UNorm, AlphaMode.Premultiplied));
                var size = new Size2(bitmap.Width, bitmap.Height);

                // Transform pixels from BGRA to RGBA
                int stride = bitmap.Width * sizeof(int);
                using (var tempStream = new DataStream(bitmap.Height * stride, true, true))
                {
                    // Lock System.Drawing.Bitmap
                    var bitmapData = bitmap.LockBits(sourceArea, System.Drawing.Imaging.ImageLockMode.ReadOnly, System.Drawing.Imaging.PixelFormat.Format32bppPArgb);

                    // Convert all pixels 
                    for (int y = 0; y < bitmap.Height; y++)
                    {
                        int offset = bitmapData.Stride * y;
                        for (int x = 0; x < bitmap.Width; x++)
                        {
                            // Not optimized 
                            byte B = Marshal.ReadByte(bitmapData.Scan0, offset++);
                            byte G = Marshal.ReadByte(bitmapData.Scan0, offset++);
                            byte R = Marshal.ReadByte(bitmapData.Scan0, offset++);
                            byte A = Marshal.ReadByte(bitmapData.Scan0, offset++);
                            int rgba = R | (G << 8) | (B << 16) | (A << 24);
                            tempStream.Write(rgba);
                        }

                    }
                    bitmap.UnlockBits(bitmapData);
                    tempStream.Position = 0;
                    return new Bitmap(renderTarget, size, tempStream, stride, bitmapProperties);
                }
            }
        }

        public static Bitmap LoadFromFile(RenderTarget renderTarget, string file)
        {
            // Loads from file using System.Drawing.Image
            using (var bitmap = (System.Drawing.Bitmap)System.Drawing.Image.FromFile(file))
            {
                var sourceArea = new System.Drawing.Rectangle(0, 0, bitmap.Width, bitmap.Height);
                var bitmapProperties = new BitmapProperties(new PixelFormat(Format.R8G8B8A8_UNorm, AlphaMode.Premultiplied));
                var size = new Size2(bitmap.Width, bitmap.Height);

                // Transform pixels from BGRA to RGBA
                int stride = bitmap.Width * sizeof(int);
                using (var tempStream = new DataStream(bitmap.Height * stride, true, true))
                {
                    // Lock System.Drawing.Bitmap
                    var bitmapData = bitmap.LockBits(sourceArea, System.Drawing.Imaging.ImageLockMode.ReadOnly, System.Drawing.Imaging.PixelFormat.Format32bppPArgb);

                    // Convert all pixels 
                    for (int y = 0; y < bitmap.Height; y++)
                    {
                        int offset = bitmapData.Stride * y;
                        for (int x = 0; x < bitmap.Width; x++)
                        {
                            // Not optimized 
                            byte B = Marshal.ReadByte(bitmapData.Scan0, offset++);
                            byte G = Marshal.ReadByte(bitmapData.Scan0, offset++);
                            byte R = Marshal.ReadByte(bitmapData.Scan0, offset++);
                            byte A = Marshal.ReadByte(bitmapData.Scan0, offset++);
                            int rgba = R | (G << 8) | (B << 16) | (A << 24);
                            tempStream.Write(rgba);
                        }

                    }
                    bitmap.UnlockBits(bitmapData);
                    tempStream.Position = 0;

                    return new Bitmap(renderTarget, size, tempStream, stride, bitmapProperties);
                }
            }
        }

        public override bool CheckClipBounds(UnE.Geometry.Vertex2D vClipTL, UnE.Geometry.Vertex2D vClipBR)
        {
            return true;
        }
        
    }
}

File: src/DXFViewerDX/DXF2DotNet/ShapeGroup.cpp
#include "StdAfx.h"
#include "ShapeGroup.h"


using namespace System::Collections;
using namespace UnE::Geometry;

BEGIN_NS(DXFDotNet)

ShapeGroup::ShapeGroup(void)
{
	m_arrShapes = gcnew ArrayList();
	m_drawType = DrawType::NONE;

	m_img = nullptr;
	m_shape = nullptr;
	m_vPos = gcnew UnE::Geometry::Vertex2D();

	m_imgSize.Width = -1.0f;
	m_imgSize.Height = -1.0f;
	m_fImageScale = 100.0f;
}

ShapeGroup::ShapeGroup(ShapeGroupOption^ option)
{
	m_arrShapes = gcnew ArrayList();

	if (option != nullptr)
	{
		m_drawType = option->DrawingType;

		m_img = option->Image;
		m_shape = option->Shape;
		m_imgSize = option->ImageSize;
	}
	else
	{
		m_drawType = DrawType::NONE;

		m_img = nullptr;
		m_shape = nullptr;
		m_imgSize.Width = -1.0f;
		m_imgSize.Height = -1.0f;
	}

	m_vPos = gcnew UnE::Geometry::Vertex2D();
	m_fImageScale = 100.0f;
}

ShapeGroup::~ShapeGroup(void)
{
}

bool ShapeGroup::GetImageSize(float% rWidth, float% rHeight)
{
	if (m_img != nullptr)
	{
		if (m_imgSize.Width < 0 || m_imgSize.Height < 0)
		{
			rWidth = m_img->Width * m_fImageScale;
			rHeight = m_img->Height * m_fImageScale;
		}
		else
		{
			rWidth = m_imgSize.Width * m_fImageScale;
			rHeight = m_imgSize.Height * m_fImageScale;
		}

		//if (m_pOwner->DownToTop())
		//	rHeight = -rHeight;

		return true;
	}

	return false;
}


// (x,y)만큼 객체를 옮긴다.
void ShapeGroup::Move(double x, double y)
{
	m_vPos->x += x;
	m_vPos->y += y;
}

Shape::ShapeType ShapeGroup::GetShapeType()
{
	return ShapeType::GROUP;
}

DXFDotNet::Shape^ ShapeGroup::Clone()
{
	ShapeGroup^ group = CreateShapeGroup();
	group->CopyFrom(this);

	group->m_vPos->x = this->m_vPos->x;
	group->m_vPos->y = this->m_vPos->y;

	group->m_drawType = this->m_drawType;

	group->m_img = this->m_img;
	group->m_shape = this->m_shape;

	for each (DXFDotNet::Shape^ shape in this->m_arrShapes)
	{
		group->AddShape(shape);
	}

	return group;
}

// Selectable이 false이면 HitTest 검사가 무조건 실패한다.
bool ShapeGroup::HitTest(double x, double y)
{
	if (!Selectable)
		return false;

	if (m_drawType == DrawType::IMAGE)
	{
		float fWidth = 0, fHeight = 0;
		if (!GetImageSize(fWidth, fHeight))
			return false;

		if (fHeight < 0.0f)
			fHeight = -fHeight;

		/*Vertex2D^ vOrigin = m_pOwner->ScreenToGlobal(0, 0);
		Vertex2D^ v2 = m_pOwner->ScreenToGlobal((int)fWidth, (int)fHeight);

		double dWidth = System::Math::Abs(v2->x - vOrigin->x);
		double dHeight = System::Math::Abs(v2->y - vOrigin->y);

		if (x >= m_vPos->x && x <= m_vPos->x + dWidth &&
			y >= m_vPos->y && y <= m_vPos->y + dHeight)
			return true;*/
		if (x >= m_vPos->x && x <= m_vPos->x + fWidth &&
			y >= m_vPos->y - fHeight && y <= m_vPos->y)
			return true;
	}
	else if (m_drawType == DrawType::SHAPE)
	{
		if (m_shape == nullptr)
			return false;

		return m_shape->HitTest(x, y);
	}

	return false;
}

void ShapeGroup::AddShape(DXFDotNet::Shape^ shape)
{
	m_arrShapes->Add(shape);
}

void ShapeGroup::RemoveShape(DXFDotNet::Shape^ shape)
{
	m_arrShapes->Remove(shape);
}

void ShapeGroup::RemoveShape(int nIndex)
{
	if (nIndex >= 0 && m_arrShapes->Count > nIndex)
		m_arrShapes->RemoveAt(nIndex);
}

void ShapeGroup::Clear()
{
	m_arrShapes->Clear();
}

int ShapeGroup::GetShapeCount()
{
	return m_arrShapes->Count;
}

DXFDotNet::Shape^ ShapeGroup::GetShape(int nIndex)
{
	if (nIndex >= 0 && m_arrShapes->Count > nIndex)
		return (DXFDotNet::Shape^)m_arrShapes[nIndex];

	return nullptr;
}

ShapeGroupOption::ShapeGroupOption()
{
	m_img = nullptr;
	m_shape = nullptr;
	m_imgSize.Width = -1.0f;
	m_imgSize.Height = -1.0f;
	m_drawType = ShapeGroup::DrawType::NONE;
}

ShapeGroupOption::ShapeGroupOption(System::Drawing::Image^ img)
{
	m_img = img;
	m_shape = nullptr;
	m_imgSize.Width = -1.0f;
	m_imgSize.Height = -1.0f;
	m_drawType = ShapeGroup::DrawType::IMAGE;
}

ShapeGroupOption::ShapeGroupOption(System::Drawing::Image^ img, float fWidth, float fHeight)
{
	m_img = img;
	m_shape = nullptr;
	m_imgSize.Width = fWidth;
	m_imgSize.Height = fHeight;
	m_drawType = ShapeGroup::DrawType::IMAGE;
}

ShapeGroupOption::ShapeGroupOption(DXFDotNet::Shape^ shape)
{
	m_img = nullptr;
	m_shape = shape;
	m_imgSize.Width = -1.0f;
	m_imgSize.Height = -1.0f;
	m_drawType = ShapeGroup::DrawType::SHAPE;
}

END_NS
